java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java

```diff
@@ -0,0 +1,450 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class SegmentGranularity
+{
+
+  // Default patterns for parsing paths.
+  final Pattern defaultPathPattern =
+      Pattern.compile(
+          "^.*[Yy]=(\\d{4})/(?:[Mm]=(\\d{2})/(?:[Dd]=(\\d{2})/(?:[Hh]=(\\d{2})/(?:[Mm]=(\\d{2})/(?:[Ss]=(\\d{2})/)?)?)?)?)?.*$"
+      );
+  final Pattern hivePathPattern =
+      Pattern.compile("^.*dt=(\\d{4})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})?)?)?)?)?)?/.*$");
+
+  public enum Formatter
+  {
+    DEFAULT,
+    HIVE,
+    LOWER_DEFAULT
+  }
+
+  public static final SegmentGranularity SECOND = SegmentGranularity.fromString("SECOND");
+  public static final SegmentGranularity MINUTE = SegmentGranularity.fromString("MINUTE");
+  public static final SegmentGranularity FIVE_MINUTE = SegmentGranularity.fromString("FIVE_MINUTE");
+  public static final SegmentGranularity TEN_MINUTE = SegmentGranularity.fromString("TEN_MINUTE");
+  public static final SegmentGranularity FIFTEEN_MINUTE = SegmentGranularity.fromString("FIFTEEN_MINUTE");
+  public static final SegmentGranularity HOUR = SegmentGranularity.fromString("HOUR");
+  public static final SegmentGranularity SIX_HOUR = SegmentGranularity.fromString("SIX_HOUR");
+  public static final SegmentGranularity DAY = SegmentGranularity.fromString("DAY");
+  public static final SegmentGranularity WEEK = SegmentGranularity.fromString("WEEK");
+  public static final SegmentGranularity MONTH = SegmentGranularity.fromString("MONTH");
+  public static final SegmentGranularity YEAR = SegmentGranularity.fromString("YEAR");
+
+  @JsonCreator
+  public static SegmentGranularity fromString(String str)
+  {
+    String name = str.toUpperCase();
+    return GranularityType.createSegmentGranularity(name);
+  }
+
+  public abstract DateTimeFormatter getFormatter(Formatter type);
+
+  public abstract DateTime increment(DateTime time);
+
+  public abstract DateTime decrement(DateTime time);
+
+  public abstract DateTime truncate(DateTime time);
+
+  public abstract DateTime toDate(String filePath, Formatter formatter);
+
+  public DateTime toDate(String filePath)
+  {
+    return toDate(filePath, Formatter.DEFAULT);
+  }
+
+  public final String toPath(DateTime time)
+  {
+    return toPath(time, "default");
+  }
+
+  private final String toPath(DateTime time, String type)
+  {
+    return toPath(time, Formatter.valueOf(type.toUpperCase()));
+  }
+
+  private final String toPath(DateTime time, Formatter type)
+  {
+    return getFormatter(type).print(time);
+  }
+
+  /**
+   * Return a granularity-sized Interval containing a particular DateTime.
+   */
+  public final Interval bucket(DateTime t)
+  {
+    DateTime start = truncate(t);
+    return new Interval(start, increment(start));
+  }
+
+  // Only to create a mapping of the granularity and all the supported file patterns
+  // namely: default, lowerDefault and hive.
+  protected enum GranularityType
```

Comment:
`granularitiesFinerThan` static method uses GranularityType. So 5, 10, 15min can't be removed easily from this enum.